sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipelineRule.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.testing;
+
+import org.apache.beam.sdk.Pipeline;
+import org.apache.beam.sdk.PipelineResult;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.runners.PipelineRunner;
+import org.apache.beam.sdk.runners.TransformHierarchy;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+/**
+ * A {@link TestRule} that verifies the following assumptions are met upon running a test that
+ * involves a {@link TestPipeline}:
+ * <ul>
+ *   <li>PAssert is not added to a pipeline that has already been run.
+ *   <li>The pipeline is run after the PAssert is added.
+ * </ul>
+ */
+public class TestPipelineRule extends TestPipeline implements TestRule {
```

Comment:
What about inlining this into `TestPipeline`? Partly, as these proliferate, I would like anyone doing this:

```java
TestPipeline p = TestPipeline.create()
```

to notice everyone else doing this:

```java
@Rule TestPipeline pipeline;
```

With the two different classes, people might think that they are different usages and both fall within best practices.

You'd have to relax some checks for tests which just run integration tests end-to-end and don't use `PAssert`, but still get their configuration from the test pipeline environment variables.